File: app/agentic/providers.ts
import { recommendations } from "../data";
import { hasLiveOracleConfig, ORACLE_AGENT_STACK, type AgenticRuntimeConfig } from "./config";
import type { EmbeddingResult, VectorSearchResult } from "./types";

const semanticCache = new Map<string, unknown>();

export class OracleCoherenceCache {
  constructor(private readonly config: AgenticRuntimeConfig) {}

  async get<T>(key: string): Promise<T | null> {
    if (this.shouldUseLiveService()) {
      throw new Error("Oracle Coherence live adapter requires service-auth implementation.");
    }

    return (semanticCache.get(key) as T | undefined) ?? null;
  }

  async set<T>(key: string, value: T): Promise<void> {
    if (this.shouldUseLiveService()) {
      throw new Error("Oracle Coherence live adapter requires service-auth implementation.");
    }

    semanticCache.set(key, value);
  }

  private shouldUseLiveService() {
    return this.config.mode === "oci" && hasLiveOracleConfig(this.config);
  }
}

export class OciCohereEmbeddingClient {
  constructor(private readonly config: AgenticRuntimeConfig) {}

  async embed(text: string): Promise<EmbeddingResult> {
    if (this.config.mode === "oci" && hasLiveOracleConfig(this.config)) {
      throw new Error("OCI Cohere Embed live adapter requires OCI request signing.");
    }

    return {
      model: ORACLE_AGENT_STACK.embeddingModel,
      dimensions: 1024,
      values: deterministicEmbedding(text, 32),
    };
  }
}

export class Oracle23AiVectorSearchClient {
  constructor(private readonly config: AgenticRuntimeConfig) {}

  async searchCourses(embedding: EmbeddingResult, query: string): Promise<VectorSearchResult[]> {
    if (this.config.mode === "oci" && hasLiveOracleConfig(this.config)) {
      throw new Error("Oracle Database 23ai Vector Search live adapter requires DB credentials.");
    }

    const normalized = query.toLowerCase();

    return recommendations
      .map((course, index) => {
        const text = `${course.title} ${course.matchReason} ${course.audience}`.toLowerCase();
        const lexicalBoost = normalized
          .split(/\W+/)
          .filter((token) => token.length > 4 && text.includes(token)).length;
        const embeddingBoost = embedding.values[index % embedding.values.length] ?? 0;

        return {
          courseId: course.courseId,
          score: Math.min(0.98, 0.72 + lexicalBoost * 0.05 + embeddingBoost * 0.03),
          reason: `Oracle 23ai vector match using ${ORACLE_AGENT_STACK.embeddingModel}`,
        };
      })
      .sort((left, right) => right.score - left.score);
  }
}

export class OciGenerativeAiCommandRClient {
  constructor(private readonly config: AgenticRuntimeConfig) {}

  async summarize(prompt: string) {
    if (this.config.mode === "oci" && hasLiveOracleConfig(this.config)) {
      throw new Error("OCI Generative AI live adapter requires OCI request signing.");
    }

    return `Cohere Command R+ mock response grounded in policy citations for: ${prompt}`;
  }
}

export class OracleAiAgentStudioClient {
  constructor(private readonly config: AgenticRuntimeConfig) {}

  async runSupervisor() {
    if (this.config.mode === "oci" && hasLiveOracleConfig(this.config)) {
      throw new Error("Oracle AI Agent Studio live adapter requires Agent Studio endpoint auth.");
    }

    return {
      supervisor: "Compliance Supervisor",
      model: ORACLE_AGENT_STACK.supervisorModel,
      tools: [
        "HCM Business Object Tool",
        "Document Tool",
        "Oracle Database 23ai Vector Search",
        "Oracle Coherence semantic cache",
      ],
    };
  }
}

export class AgentStudioEvaluationClient {
  constructor(private readonly config: AgenticRuntimeConfig) {}

  async score(citationCount: number, recordCount: number) {
    if (this.config.mode === "oci" && hasLiveOracleConfig(this.config)) {
      throw new Error("Agent Studio evaluation live adapter requires evaluation API binding.");
    }

    const citationCoverage = recordCount === 0 ? 1 : Math.min(1, citationCount / (recordCount * 2));
    const faithfulness = citationCoverage >= 0.9 ? 0.96 : 0.88;
    const answerRelevance = recordCount > 0 ? 0.93 : 0.8;

    return {
      framework: ORACLE_AGENT_STACK.evaluationProvider,
      dataset: "50-case compliance golden set",
      faithfulness,
      answerRelevance,
      citationCoverage,
      deploymentGate: citationCoverage >= 0.85 ? ("pass" as const) : ("review" as const),
    };
  }
}

function deterministicEmbedding(text: string, dimensions: number) {
  const values: number[] = [];
  let hash = 2166136261;

  for (let index = 0; index < text.length; index += 1) {
    hash ^= text.charCodeAt(index);
    hash = Math.imul(hash, 16777619);
  }

  for (let index = 0; index < dimensions; index += 1) {
    hash ^= index + 0x9e3779b9;
    hash = Math.imul(hash, 16777619);
    values.push(((hash >>> 0) % 1000) / 1000);
  }

  return values;
}

File: backend/agentic/providers.py
from __future__ import annotations

from dataclasses import dataclass

from agentic.config import ORACLE_AGENT_STACK, RuntimeConfig
from agentic.data import RECOMMENDATIONS


_SEMANTIC_CACHE: dict[str, dict[str, object]] = {}


@dataclass(frozen=True)
class EmbeddingResult:
    model: str
    dimensions: int
    values: list[float]


@dataclass(frozen=True)
class VectorSearchResult:
    course_id: str
    score: float
    reason: str


class OracleCoherenceCache:
    def __init__(self, config: RuntimeConfig):
        self.config = config

    def get(self, key: str) -> dict[str, object] | None:
        if self._should_use_live_service():
            raise NotImplementedError("Oracle Coherence live adapter requires service auth.")

        return _SEMANTIC_CACHE.get(key)

    def set(self, key: str, value: dict[str, object]) -> None:
        if self._should_use_live_service():
            raise NotImplementedError("Oracle Coherence live adapter requires service auth.")

        _SEMANTIC_CACHE[key] = value

    def _should_use_live_service(self) -> bool:
        return self.config.mode == "oci" and self.config.has_live_oracle_config


class OciCohereEmbeddingClient:
    def __init__(self, config: RuntimeConfig):
        self.config = config

    def embed(self, text: str) -> EmbeddingResult:
        if self.config.mode == "oci" and self.config.has_live_oracle_config:
            raise NotImplementedError("OCI Cohere Embed live adapter requires OCI request signing.")

        return EmbeddingResult(
            model=ORACLE_AGENT_STACK["embedding_model"],
            dimensions=1024,
            values=_deterministic_embedding(text, 32),
        )


class Oracle23AiVectorSearchClient:
    def __init__(self, config: RuntimeConfig):
        self.config = config

    def search_courses(self, embedding: EmbeddingResult, query: str) -> list[VectorSearchResult]:
        if self.config.mode == "oci" and self.config.has_live_oracle_config:
            raise NotImplementedError("Oracle Database 23ai live adapter requires DB credentials.")

        normalized = query.lower()
        query_tokens = [token for token in _tokenize(normalized) if len(token) > 4]
        matches: list[VectorSearchResult] = []

        for index, course in enumerate(RECOMMENDATIONS):
            text = f"{course.title} {course.match_reason} {course.audience}".lower()
            lexical_boost = sum(1 for token in query_tokens if token in text)
            embedding_boost = embedding.values[index % len(embedding.values)]
            score = min(0.98, 0.72 + lexical_boost * 0.05 + embedding_boost * 0.03)
            matches.append(
                VectorSearchResult(
                    course_id=course.course_id,
                    score=score,
                    reason=f"Oracle 23ai vector match using {ORACLE_AGENT_STACK['embedding_model']}",
                )
            )

        return sorted(matches, key=lambda item: item.score, reverse=True)


class OciGenerativeAiCommandRClient:
    def __init__(self, config: RuntimeConfig):
        self.config = config

    def summarize(self, prompt: str) -> str:
        if self.config.mode == "oci" and self.config.has_live_oracle_config:
            raise NotImplementedError("OCI Generative AI live adapter requires OCI request signing.")

        return f"Cohere Command R+ mock response grounded in policy citations for: {prompt}"


class OracleAiAgentStudioClient:
    def __init__(self, config: RuntimeConfig):
        self.config = config

    def run_supervisor(self) -> dict[str, object]:
        if self.config.mode == "oci" and self.config.has_live_oracle_config:
            raise NotImplementedError("Oracle AI Agent Studio live adapter requires endpoint auth.")

        return {
            "supervisor": "Compliance Supervisor",
            "model": ORACLE_AGENT_STACK["supervisor_model"],
            "tools": [
                "HCM Business Object Tool",
                "Document Tool",
                "Oracle Database 23ai Vector Search",
                "Oracle Coherence semantic cache",
            ],
        }


class AgentStudioEvaluationClient:
    def __init__(self, config: RuntimeConfig):
        self.config = config

    def score(self, citation_count: int, record_count: int) -> dict[str, object]:
        if self.config.mode == "oci" and self.config.has_live_oracle_config:
            raise NotImplementedError("Agent Studio evaluation live adapter requires API binding.")

        citation_coverage = 1 if record_count == 0 else min(1, citation_count / (record_count * 2))
        faithfulness = 0.96 if citation_coverage >= 0.9 else 0.88
        answer_relevance = 0.93 if record_count > 0 else 0.8

        return {
            "framework": ORACLE_AGENT_STACK["evaluation_provider"],
            "dataset": "50-case compliance golden set",
            "faithfulness": faithfulness,
            "answerRelevance": answer_relevance,
            "citationCoverage": citation_coverage,
            "deploymentGate": "pass" if citation_coverage >= 0.85 else "review",
        }


def _deterministic_embedding(text: str, dimensions: int) -> list[float]:
    hash_value = 2166136261

    for character in text:
        hash_value ^= ord(character)
        hash_value = (hash_value * 16777619) & 0xFFFFFFFF

    values: list[float] = []

    for index in range(dimensions):
        hash_value ^= index + 0x9E3779B9
        hash_value = (hash_value * 16777619) & 0xFFFFFFFF
        values.append((hash_value % 1000) / 1000)

    return values


def _tokenize(value: str) -> list[str]:
    token = ""
    tokens: list[str] = []

    for character in value:
        if character.isalnum():
            token += character
        elif token:
            tokens.append(token)
            token = ""

    if token:
        tokens.append(token)

    return tokens
